frontend/src/metabase-lib/lib/queries/utils/native-query.ts

```diff
@@ -0,0 +1,37 @@
+import slugg from "slugg";
```

Comment:
@calherries @daltojohnso, this is almost RFC-like.

The dir structure and file naming we have been using.

What if we used:

`frontend/src/lib/queries/NativeQuery/NativeQuery.ts` (the Class)
and for this file here
`frontend/src/lib/queries/NativeQuery/names.ts` (or something better than `names.ts`, I'm more concerned with the dir structure tbh.

Why?

1. `metabase-lib/lib` serves no purpose being 2 dirs.
2. `metabase-lib` as a name adds little...we know this is metabase we're working on.
3. then we put all files for a given lib Class in one directory only

Does this make sense to you?

Reply:
Yes, it definitely makes sense to rethink this, but I see this as mostly orthogonal to the task I'm working on. Can I just put these functions at the bottom of `frontend/src/metabase-lib/lib/queries/NativeQuery.ts` for now, and we can move them later? 

At least I will leave the functions pure, and outside the class, so refactoring is as simple as changing the file name.

Reply:
Imo it's better to keep them here than in NativeQuery.ts 👍

Reply:
At the same time (and this is not a comment on anyone, rather on what emerges collectively), as we focus tightly on the issue at hand, and almost always move to another issue right after solving the previous immediate issue (and individually both decisions are justifiable), the patterns that emerge are suboptimal and over several years nits become a serious lice infestation :)

In other words, somehow we should make sure to improve the architecture as we go along.

Reply:
Alright, then I'll need to lean heavily on the frontend team's guidance, because I don't have the context to decide the directory structure without your guidance.

For my understanding of the codebase, I've written down all the files where we have a concept of a "query":

- `metabase/lib/query.js` has `createQuery` and a few higher level functions written in a pure functional style.
- `metabase/lib/query/query.js` defines a pure functional API for a "query" (untyped though), aggregating all the subcomponents in the `metabase/lib/query/...` directory, e.g. `getAggregations`.
- `metabase-lib/lib/queries/Query.ts` defines the `Query` class, basically a wrapper for `DatasetQuery`, but references a question and metadata.
- 
- `metabase-lib/lib/queries/NativeQuery.ts` and `metabase-lib/lib/queries/StructuredQuery.ts` defines constructors for `NativeQuery` and `StructuredQuery` classes. `StructuredQuery` appears to wrap a lot of the methods defined by `metabase/lib/query.js`, but `NativeQuery` doesn't have a similar namespace it depends on.
- `metabase-types/types/Card.ts` defines `StructuredDatasetQuery`, `NativeDatasetQuery`, and their supertype `DatasetQuery`, which wrap `NativeQuery` and `StructuredQuery` respectively with extra info, like a database ID
- `metabase-types/types/Query.ts` defines types for `NativeQuery` and `StructuredQuery`, but not `Query`. Note `StructuredQuery` type is distinct from the `StructuredQuery` class defined in `metabase-lib/lib/queries/StructuredQuery.ts`, and the `NativeQuery` type is distinct from the `NativeQuery` class defined in `metabase-lib/lib/queries/NativeQuery.ts`

But based on this I think the most sensible to do is to ignore the problem, and move on :p

Just joking. 

But it's definitely above my FE abilities right now to reason about where these two little functions should live. I think I'd rather leave that up to the big boys

Reply:
@gusaiani I moved these functions back to `metabase-lib/lib/queries/NativeQuery.ts` for now, as I wanted them all in one place as I figured out how they interacted with the existing helper functions in `metabase-lib/lib/queries/NativeQuery.ts`. I've refactored `NativeQuery.ts` so that all these functions are at least in one block, which will make moving them somewhere else easier. See this:
https://github.com/metabase/metabase/blob/d156890b17f4f601cb00f63853d24a89d093d1f6/frontend/src/metabase-lib/lib/queries/NativeQuery.ts#L49-L54

Reply:
Gotcha.

Reply:
Replying to your comment [here](https://github.com/metabase/metabase/pull/25319/files/44d426e48c7620f12b1e93eca9a840c6a2e1767b#r972492191) on this thread to not lose context 


> This new version is beautiful.
> 
> I may have expressed myself badly in today's comment, because imo this new function should live in that extracted file. Not in this class file.
> 
> If you're done tweaking this, I'll understand.


To be clear I'm not done tweaking this, as I'd like to do it right. Moving these functions back was a refactoring step towards getting all functions related to query text and tags in one place. I felt it more important to have all closely coupled functions in one file (even if big) than have the logic split into different files. Now they're (mostly) all together in one block, I just need to know where to put them.